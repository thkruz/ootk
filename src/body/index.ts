export { Celestial } from './Celestial';
export { Earth } from './Earth';
export { Moon } from './Moon';
export { Sun } from './Sun';
