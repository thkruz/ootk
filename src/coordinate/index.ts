export * from './Hill.js';
export * from './ClassicalElements.js';
export * from './EquinoctialElements.js';
export * from './FormatTle.js';
export * from './Geodetic.js';
export * from './ITRF.js';
export * from './J2000.js';
export * from './RelativeState.js';
export * from './RIC.js';
export * from './StateVector.js';
export * from './TEME.js';
export * from './Tle.js';
