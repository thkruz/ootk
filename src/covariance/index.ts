export * from './CovarianceSample.js';
export * from './StateCovariance.js';
