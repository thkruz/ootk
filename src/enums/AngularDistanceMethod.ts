/**
 * @author @thkruz Theodore Kruczek
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

/** Enumeration representing different methods for calculating angular distance. */
export enum AngularDistanceMethod {
  Cosine,
  Haversine,
}
