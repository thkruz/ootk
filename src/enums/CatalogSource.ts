export enum CatalogSource {
  USSF = 'USSF',
  CELESTRAK = 'Celestrak',
  UNIV_OF_MICH = 'University of Michigan',
  CALPOLY = 'CalPoly',
  NUSPACE = 'NuSpace',
  VIMPEL = 'JSC Vimpel',
  TLE_TXT = 'TLE.txt',
  EXTRA_JSON = 'extra.json',
  SATNOGS = 'SatNOGS',
}
