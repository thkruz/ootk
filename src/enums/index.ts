export { CatalogSource } from './CatalogSource.js';
export { CommLink } from './CommLink.js';
export { OrbitRegime } from './OrbitRegime.js';
export { Sgp4OpsMode } from './Sgp4OpsMode.js';
export { AngularDiameterMethod } from './AngularDiameterMethod.js';
export { AngularDistanceMethod } from './AngularDistanceMethod.js';
export { PassType } from './PassType.js';
