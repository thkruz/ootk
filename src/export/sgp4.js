import { Sgp4Js } from '../sgp4.js';
// import { Sgp4JsUtils } from '../sgp4-utils.js';

window.Sgp4Js = Sgp4Js;
// window.Sgp4JsUtils = Sgp4JsUtils;
