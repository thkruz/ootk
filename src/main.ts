/**
 * @author @thkruz Theodore Kruczek
 * @description Orbital Object ToolKit (ootk) is a collection of tools for working
 * with satellites and other orbital objects.
 * @license AGPL-3.0-or-later
 * @copyright (c) 2020-2024 Theodore Kruczek
 *
 * Many of the classes are based off of the work of @david-rc-dayton and his
 * Pious Squid library (https://github.com/david-rc-dayton/pious_squid).
 *
 * Orbital Object ToolKit is free software: you can redistribute it and/or modify it under the
 * terms of the GNU Affero General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * Orbital Object ToolKit is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License along with
 * Orbital Object ToolKit. If not, see <http://www.gnu.org/licenses/>.
 */

export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './time/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
export * from './operations/operations.js';
export { BaseObject, Sensor } from './objects/index.js';
export { Earth } from './body/index.js';
export * from './coordinate/index.js';
export * from './observation/index.js';
export * from './data/DataHandler.js';
export * from './sgp4/index.js';
export * from './objects/index.js';
export * from './interpolator/index.js';
export * from './force/index.js';
export * from './maneuver/index.js';
export * from './optimize/index.js';
export * from './orbit_determination/index.js';
export * from './body/index.js';
export * from './enums/index.js';
export * from './types/types.js';
export * from './interfaces/index.js';
export * from './coordinate/index.js';
export * from './operations/index.js';
export * from './force/index.js';
export * from './propagator/index.js';
export * from './transforms/index.js';
export * from './utils/index.js';
