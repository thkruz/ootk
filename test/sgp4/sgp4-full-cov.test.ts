import { Sgp4OpsMode } from '../../src/enums/Sgp4OpsMode';
import { EciVec3, KilometersPerSecond, Sgp4, Vec3Flat } from '../../src/main';
import { Sgp4GravConstants } from '../../src/sgp4/sgp4';
import { sgp4FullCov } from './sgp4-full-cov';
import { sgp4FullCovFail } from './sgp4-full-cov-fail';

describe('Verification TLE Data in Appendix D of Revisiting Spacetrack Report #3: Rev 1', () => {
  sgp4FullCov.forEach((sgp4DataItem) => {
    // Fetching satellite record from TLE lines
    const satrec = Sgp4.createSatrec(
      sgp4DataItem.tleLine1,
      sgp4DataItem.tleLine2,
      Sgp4GravConstants.wgs72,
      Sgp4OpsMode.AFSPC,
    );

    if ((sgp4DataItem as unknown as { error: true }).error) {
      test(`if ${sgp4DataItem.description} fails`, () => {
        expect(satrec.error).toEqual((sgp4DataItem as unknown as { error: true }).error);
      });
    } else {
      test(`if ${sgp4DataItem.description} passes`, () => {
        sgp4DataItem.results.forEach((expected) => {
          const sgp4Result = Sgp4.propagate(satrec, expected.time);

          expect(sgp4Result).not.toBe(false);
          const position = sgp4Result?.position as EciVec3;
          const velocity = sgp4Result?.velocity as EciVec3<KilometersPerSecond>;

          expect(position.x).toBeCloseTo(expected.position.x);
          expect(position.y).toBeCloseTo(expected.position.y);
          expect(position.z).toBeCloseTo(expected.position.z);
          expect(velocity.x).toBeCloseTo(expected.velocity.x);
          expect(velocity.y).toBeCloseTo(expected.velocity.y);
          expect(velocity.z).toBeCloseTo(expected.velocity.z);
        });
      });
    }
  });
});

describe('Verify getgravconst options', () => {
  const line1 = '1 00005U 58002B   00179.78495062  .00000023  00000-0  28098-4 0  4753';
  const line2 = '2 00005  34.2682 348.7242 1859667 331.7664  19.3264 10.82419157413667';

  test('if wgs72old can be selected', () => {
    const satrec = Sgp4.createSatrec(line1, line2, Sgp4GravConstants.wgs72old, Sgp4OpsMode.AFSPC);

    expect(satrec.xmcof).toEqual(-1.8859361255715234e-11);
  });
  test('if wgs84 can be selected', () => {
    const satrec = Sgp4.createSatrec(line1, line2, Sgp4GravConstants.wgs84, Sgp4OpsMode.AFSPC);

    expect(satrec.xmcof).toEqual(-1.8859472970032445e-11);
  });
  test('if other gravconst values cause an error', () => {
    expect(() => Sgp4.createSatrec(line1, line2, 'wgs96' as Sgp4GravConstants, Sgp4OpsMode.AFSPC)).toThrow(
      new Error('unknown gravity option wgs96'),
    );
  });
});

describe('Verification of Fail Cases', () => {
  sgp4FullCovFail.forEach((sgp4DataItem) => {
    // Fetching satellite record from TLE lines
    const satrec = Sgp4.createSatrec(sgp4DataItem.tleLine1, sgp4DataItem.tleLine2);

    test(`if ${sgp4DataItem.description} fails`, () => {
      expect(satrec.error).toEqual(sgp4DataItem.error);
    });
  });
});

describe('Test vector equations in SGP4', () => {
  test('if invjday calculates date', () => {
    expect(Sgp4.invjday(2450000, 0)).toEqual({
      day: 9,
      hr: 12,
      min: 0,
      mon: 10,
      sec: 0,
      year: 1995,
    });
    expect(Sgp4.jday(1995, 10, 9, 12, 0, 0)).toEqual({
      jd: 2449999.5,
      jdFrac: 0.5,
    });

    const jday1 = { jd: 2450000, jdFrac: 0 };
    const epoch1 = jday1.jd + jday1.jdFrac;
    const jday2 = Sgp4.jday(1995, 10, 9, 12, 0, 0);
    const epoch2 = jday2.jd + jday2.jdFrac;

    expect(epoch1).toEqual(epoch2);
  });
});

describe('Ensure bstar and ndot account for leading zeros', () => {
  test('if bstar reads in leading zero without exponent', () => {
    const line1 = '1 00005U 58002B   00179.78495062  .00000023  00000-0  00098-0 0  4753';
    const line2 = '2 00005  34.2682 348.7242 1859667 331.7664  19.3264 10.82419157413667';
    const satrec = Sgp4.createSatrec(line1, line2, Sgp4GravConstants.wgs72, Sgp4OpsMode.IMPROVED);

    expect(satrec.bstar).toEqual(0.00098);
  });
  test('if bstar reads in leading zero when doing exponents', () => {
    const line1 = '1 00005U 58002B   00179.78495062  .00000023  00000-0  00098-5 0  4753';
    const line2 = '2 00005  34.2682 348.7242 1859667 331.7664  19.3264 10.82419157413667';
    const satrec = Sgp4.createSatrec(line1, line2, Sgp4GravConstants.wgs72, Sgp4OpsMode.IMPROVED);

    expect(satrec.bstar).toEqual(0.0000000098);
  });
  test('if bstar reads in negative number', () => {
    const line1 = '1 00005U 58002B   00179.78495062  .00000023  00000-0 -00098-5 0  4753';
    const line2 = '2 00005  34.2682 348.7242 1859667 331.7664  19.3264 10.82419157413667';
    const satrec = Sgp4.createSatrec(line1, line2, Sgp4GravConstants.wgs72, Sgp4OpsMode.IMPROVED);

    expect(satrec.bstar).toEqual(-0.0000000098);
  });
  test('if ndot reads in leading zero', () => {
    const line1 = '1 00005U 58002B   00179.78495062  .10000000  00023-0  12398-0 0  4753';
    const line2 = '2 00005  34.2682 348.7242 1859667 331.7664  19.3264 10.82419157413667';
    const satrec = Sgp4.createSatrec(line1, line2, Sgp4GravConstants.wgs72, Sgp4OpsMode.IMPROVED);

    expect(satrec.nddot).toEqual(0.00023);
  });
  test('if ndot reads in leading zero and applies exponents', () => {
    const line1 = '1 00005U 58002B   00179.78495062  .10000000  00023-4  12398-0 0  4753';
    const line2 = '2 00005  34.2682 348.7242 1859667 331.7664  19.3264 10.82419157413667';
    const satrec = Sgp4.createSatrec(line1, line2, Sgp4GravConstants.wgs72, Sgp4OpsMode.IMPROVED);

    expect(satrec.nddot).toEqual(0.000000023);
  });
  test('if ndot reads in negative value with leading zeroes', () => {
    const line1 = '1 00005U 58002B   00179.78495062 -.10000000 -00023-0  12398-5 0  4753';
    const line2 = '2 00005  34.2682 348.7242 1859667 331.7664  19.3264 10.82419157413667';
    const satrec = Sgp4.createSatrec(line1, line2, Sgp4GravConstants.wgs72, Sgp4OpsMode.IMPROVED);

    expect(satrec.nddot).toEqual(-0.00023);
  });

  test('rv2coe', () => {
    const r = [6524.834, 6862.875, 6448.296] as Vec3Flat;
    const v = [4.901327, 5.533756, -1.976341] as Vec3Flat;

    expect(Sgp4.rv2coe(r, v, 398600.8)).toMatchSnapshot();
  });
});
