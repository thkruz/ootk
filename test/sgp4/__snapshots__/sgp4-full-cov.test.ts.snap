// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ensure bstar and ndot account for leading zeros rv2coe 1`] = `
Object {
  "a": 36127.16533103146,
  "arglat": 999999.1,
  "argp": 0.9317417321423879,
  "ecc": 0.8328526866648804,
  "incl": 1.5336055626394494,
  "lonper": 999999.1,
  "m": 0.13272880776065465,
  "nu": 1.6115535789428714,
  "omega": 3.9775750028016947,
  "p": 11067.78839665728,
  "truelon": 999999.1,
}
`;
